### PR TITLE
Eventing-Kafka 0.26 KafkaChannel post job update

### PR DESCRIPTION
**Pull request by matzew**
Adding the post-install job for the channel, a follow-up PR for https://github.com/openshift-knative/serverless-operator/pull/1286

### CODE REVIEW

**Comment by matzew** [comment]
>         --- FAIL: TestServerlessUpgrade/VerifyContinualTests/AutoscaleSustainingTest (0.36s)


/retest

File: knative-operator/hack/009-generated-job-name.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/knative-operator/deploy/resources/knativekafka/3-channel-post-install.yaml b/knative-operator/deploy/resources/knativekafka/3-channel-post-install.yaml
+index c316076c..8258531c 100644
+--- a/knative-operator/deploy/resources/knativekafka/3-channel-post-install.yaml
++++ b/knative-operator/deploy/resources/knativekafka/3-channel-post-install.yaml
+@@ -162,7 +162,7 @@ roleRef:
+ apiVersion: batch/v1
+ kind: Job
+ metadata:
+-  name: eventing-kafka-v0.26-channel-post-install-job
++  name: v0.26-eventing-kafka-channel-post-install-job
```
Comment (matzew) [inline knative-operator/hack/009-generated-job-name.patch:10]: Using the 'old' semantics for "update" images.

Starting with 1.21 (or upstream 0.27/1.0) we will have `generatedName`...  see the `update-manifest` for the actual comment

**Comment by matzew** [comment]
/hold

looks like something on the teardown of the upgrade test

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go [hunk 2]
```diff
@@ -61,16 +61,22 @@ func newReconciler(mgr manager.Manager) (*ReconcileKnativeKafka, error) {
...
 
 	reconcileKnativeKafka := ReconcileKnativeKafka{
-		client:                  mgr.GetClient(),
-		scheme:                  mgr.GetScheme(),
-		rawKafkaChannelManifest: kafkaChannelManifest,
-		rawKafkaSourceManifest:  kafkaSourceManifest,
+		client:                             mgr.GetClient(),
+		scheme:                             mgr.GetScheme(),
+		rawKafkaChannelManifest:            kafkaChannelManifest,
+		rawKafkaChannelPostInstallManifest: kafkaChannelPostInstallManifest,
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:78]: Why do we need a specific post-install manifest? Could we not add it to the kafkaChannelManifest directly?

File: knative-operator/hack/009-generated-job-name.patch (hunk 1 above)
Comment (markusthoemmes) [reply knative-operator/hack/009-generated-job-name.patch:10]: Should we patch towards the "future" version, i.e. make this patch to the `generatedName`: `storage-version-migration-eventing-kafka-channel-` as done in https://github.com/knative-sandbox/eventing-kafka/blob/79f02ebad7dc2fa9117236d85d06fb83b36c2fa0/config/channel/post-install/storage-version-migrator.yaml#L18?

We'd then not need to change the csv.sh etc. again and just drop the patch in 1.0.

**Comment by markusthoemmes** [comment]
Towards the error: The job pods are still around, the new one is crashlooping because of

```
{"level":"info","ts":1636066638.7470193,"logger":"fallback","caller":"client/config.go:289","msg":"Built Sarama config: &{Admin:{Retry:{Max:5 Backoff:100ms} Timeout:3s} Net:{MaxOpenRequests:5 DialTimeout:30s ReadTimeout:30s WriteTimeout:30s TLS:{Enable:false Config:<nil>} SASL:{Enable:false Mechanism: Version:0 Handshake:true AuthIdentity: User: Password: SCRAMAuthzID: SCRAMClientGeneratorFunc:<nil> TokenProvider:<nil> GSSAPI:{AuthType:0 KeyTabPath: KerberosConfigPath: ServiceName: Username: Password: Realm: DisablePAFXFAST:false}} KeepAlive:0s LocalAddr:<nil> Proxy:{Enable:false Dialer:<nil>}} Metadata:{Retry:{Max:3 Backoff:250ms BackoffFunc:<nil>} RefreshFrequency:10m0s Full:true Timeout:0s} Producer:{MaxMessageBytes:1000000 RequiredAcks:-1 Timeout:10s Compression:none CompressionLevel:-1000 Partitioner:0x7e6500 Idempotent:false Return:{Successes:true Errors:true} Flush:{Bytes:0 Messages:0 Frequency:0s MaxMessages:0} Retry:{Max:3 Backoff:100ms BackoffFunc:<nil>} Interceptors:[]} Consumer:{Group:{Session:{Timeout:10s} Heartbeat:{Interval:3s} Rebalance:{Strategy:0x2a87400 Timeout:1m0s Retry:{Max:4 Backoff:2s}} Member:{UserData:[]}} Retry:{Backoff:2s BackoffFunc:<nil>} Fetch:{Min:1 Default:1048576 Max:0} MaxWaitTime:250ms MaxProcessingTime:100ms Return:{Errors:true} Offsets:{CommitInterval:0s AutoCommit:{Enable:true Interval:1s} Initial:-1 Retention:0s Retry:{Max:3}} IsolationLevel:0 Interceptors:[]} ClientID:eventing-kafka-v0.26-post-install-job RackID: ChannelBufferSize:256 Version:1.0.0 MetricRegistry:0xc00004d900}"}
{"level":"fatal","ts":1636066638.7470658,"logger":"fallback","caller":"post-install/main.go:83","msg":"Loaded Invalid Configuration Settings","stacktrace":"main.main\n\t/go/src/knative.dev/eventing-kafka/cmd/channel/post-install/main.go:83\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225"}
```

**Comment by matzew** [comment]
yeah the image/code does require the "new" config: 
https://github.com/knative-sandbox/eventing-kafka/blob/release-0.26/cmd/channel/post-install/main.go#L82

We still patch for the old one, that is still supported  - will look at this

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 2 above)
Comment (matzew) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:87]: You mean `.Append()` to the `kafkaChannelManifest`? I could, I guess - I was thinking to keep them separated :man_shrugging:

File: knative-operator/hack/009-generated-job-name.patch (hunk 1 above)
Comment (matzew) [reply knative-operator/hack/009-generated-job-name.patch:10]: yeah, but next release (1.0) we have a new post-install for the source too. Than they all are `generatedName`, I did that upstream - and than I think I am generally touching them, so will do it than - for all

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go (hunk 2 above)
Comment (matzew) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:87]: I like the idea of a folder - Are you OK, I do refactor this afterwards?

(as a prep. for the 1.0 based post jobs for source/channel - and soon we have broker too - so channel makes sense)

Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:87]: Personally, I would prefactor it, but I'm fine either way.

**Comment by matzew** [comment]
/hold 
will rebase after https://github.com/openshift-knative/serverless-operator/pull/1306

File: knative-operator/hack/009-generated-job-name.patch (hunk 1 above)
Comment (markusthoemmes) [reply knative-operator/hack/009-generated-job-name.patch:10]: alrighty

**Comment by matzew** [comment]
/unhold

was rebased

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

**Pull request merged**